lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -413,14 +268,64 @@ struct AssertOpConversion
         return failure();
       }
     }
+#ifdef USE_ROCM
+    llAssertHIP(op, condition, adaptor.getMessage(), adaptor.getFile(),
+                adaptor.getFunc(), adaptor.getLine(), rewriter);
+#else
     llAssert(op, condition, adaptor.getMessage(), adaptor.getFile(),
              adaptor.getFunc(), adaptor.getLine(), rewriter);
+#endif
     rewriter.eraseOp(op);
     return success();
   }
 
   // op: the op at which the assert is inserted. Unlike printf, we need to
   // know about the op to split the block.
+#ifdef USE_ROCM
+  void llAssertHIP(Operation *op, Value condition, StringRef message,
+                          StringRef file, StringRef func, int line,
+                          ConversionPatternRewriter &rewriter) const {
+    ConversionPatternRewriter::InsertionGuard guard(rewriter);
+    auto ctx = rewriter.getContext();
+    auto loc = op->getLoc();
+
+    // #block1
+    // if (condition) {
+    //   #block2
+    //   print(message);
+    //   halt;
+    // }
+    // #block3
+    Block *prevBlock = op->getBlock();
+    Block *ifBlock = rewriter.splitBlock(prevBlock, op->getIterator());
+    rewriter.setInsertionPointToStart(ifBlock);
+
+    SmallString<256> tmpBuf;
+    message =
+        llvm::Twine("Assertion failed: " + message + ", File: " + file +
+                    ", Function: " + func + ", Line: " + llvm::Twine(line))
+            .toStringRef(tmpBuf);
+
+    // Print assert message.
+    llPrintfHIP(loc, op->getParentOfType<mlir::ModuleOp>(), message,
+                ValueRange(), rewriter, /*stderr*/ true);
+
+    // Perform the trap.
+    GCNBuilder BuilderTrap;
+    // TODO: s_trap instruction doesn't work here.
+    BuilderTrap.create<>("s_endpgm")->operator()();
+    BuilderTrap.launch(rewriter, loc, void_ty(ctx));
```

Comment:
Is it possible to use something like `llvm.unreachable` instead of inline assembler here?

Reply:
Nevermind, probably this is not how unreachable is mentioned to be used

Reply:
Oh, these intrinsics was added last week actually )) Will try to use them.

Reply:
Unfortunately, both `LLVM::Trap` and `LLVM::DebugTrap` instructions are not stable on my HW. So I decided to leave `s_endpgm` here as it is suggested as a workaround here: https://llvm.org/docs/AMDGPUUsage.html#id52